`#[local_ref]` does not always work when adding widgets to containers

Consider this example:


```rust
let gtk_box = gtk::Box::default();
let gtk_box_ref = &gtk_box;

relm4::view! {
    #[local_ref]
    gtk_box_ref {
        set_orientation: gtk::Orientation::Vertical,

        gtk::Button {
            set_label: "label",
        }
    }
}
```

Here I use `#[local_ref]` attribute to have `gtk_box_ref` in `view!`. I intend to append a button to the `gtk_box`.

However, it doesn't compile:

```rust
error[E0277]: the trait bound `&relm4::gtk4::Box: RelmSetChildExt` is not satisfied
  --> examples/simple.rs:66:13
   |
66 |             gtk_box_ref {
   |             ^^^^^^^^^^^ the trait `RelmSetChildExt` is not implemented for `&relm4::gtk4::Box`
   |
...
= note: required because of the requirements on the impl of `RelmContainerExt` for `&relm4::gtk4::Box`
```

If we expand the macro, here's the line that's causing problems:

```rust
relm4::RelmContainerExt::container_add(&gtk_box_ref, &_gtk_button_4);
```

It's a reference to a reference. It seems like the compiler is having a hard time to deref `&&gtk::Box` into `&gtk::Box`.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -5,6 +5,11 @@
 + core: `add_action` of `RelmActionGroup` now takes a reference to a `RelmAction` as a parameter
 + macros: `parse_with_path`, `update_stream`, `inject_view_code` and `generate_tokens` take references for some of their parameters
 
+### Fixed
+
++ macros: Fix usage of RelmContainerExt with local_ref attribute
++ macros: Report RelmContainerExt error at the correct span
+
 ## 0.5.0-beta.3 - 2022-9-28
 
 ### Added

--- a/relm4-macros/src/lib.rs
+++ b/relm4-macros/src/lib.rs
@@ -212,7 +212,7 @@ pub fn component(attributes: TokenStream, input: TokenStream) -> TokenStream {
 ///     SendFront(DynamicIndex),
 /// }
 ///
-/// #[relm4::factory(pub)]
+/// #[relm4_macros::factory(pub)]
 /// impl FactoryComponent for Counter {
 ///     type CommandOutput = ();
 ///     type Init = u8;

--- a/relm4-macros/src/token_streams.rs
+++ b/relm4-macros/src/token_streams.rs
@@ -110,6 +110,11 @@ impl Widget {
             self.init_stream(&mut streams.init);
         }
 
+        #[cfg(feature = "relm4")]
+        streams.assign.extend(quote::quote! {
+            use relm4::RelmContainerExt as _;
+        });
+
         self.error_stream(&mut streams.error);
         self.start_assign_stream(&mut streams.assign);
         self.init_conditional_init_stream(&mut streams.assign, model_name);

--- a/relm4-macros/src/widgets/gen/util/property_name.rs
+++ b/relm4-macros/src/widgets/gen/util/property_name.rs
@@ -11,18 +11,16 @@ impl PropertyName {
                 quote! { #w_name.#ident }
             }
             PropertyName::Path(path) => path.to_token_stream(),
-            PropertyName::RelmContainerExtAssign => {
-                quote_spanned! { w_name.span() => relm4::RelmContainerExt::container_add }
+            PropertyName::RelmContainerExtAssign(span) => {
+                quote_spanned! { *span => #w_name.container_add }
             }
         }
     }
 
     pub(crate) fn assign_args_stream(&self, w_name: &Ident) -> Option<TokenStream2> {
         match self {
-            PropertyName::Ident(_) => None,
-            PropertyName::Path(_) | PropertyName::RelmContainerExtAssign => {
-                Some(quote_spanned! { w_name.span() => & #w_name, })
-            }
+            PropertyName::RelmContainerExtAssign(_) | PropertyName::Ident(_) => None,
+            PropertyName::Path(_) => Some(quote_spanned! { w_name.span() => & #w_name, }),
         }
     }
 }

--- a/relm4-macros/src/widgets/mod.rs
+++ b/relm4-macros/src/widgets/mod.rs
@@ -73,7 +73,7 @@ struct ClosureSignalHandler {
 enum PropertyName {
     Ident(Ident),
     Path(Path),
-    RelmContainerExtAssign,
+    RelmContainerExtAssign(Span2),
 }
 
 struct Property {

--- a/relm4-macros/src/widgets/parse/property.rs
+++ b/relm4-macros/src/widgets/parse/property.rs
@@ -32,7 +32,7 @@ impl Property {
         // parse `if something { WIDGET } else { WIDGET}` or a similar match expression.
         if input.peek(Token![if]) || input.peek(Token![match]) {
             return Ok(Property {
-                name: PropertyName::RelmContainerExtAssign,
+                name: PropertyName::RelmContainerExtAssign(input.span()),
                 ty: PropertyType::ConditionalWidget(ConditionalWidget::parse_with_name(
                     input,
                     None,
@@ -47,11 +47,12 @@ impl Property {
 
         // `gtk::Box { ... }`, `data.init_widget() -> gtk::Button { ... }` or `gtk::Box,`
         if input.peek(token::Brace) || input.peek(Token![->]) || input.peek(Token![,]) {
+            let span = func.span();
             let ty =
                 PropertyType::Widget(Widget::parse_for_container_ext(input, func, attributes)?);
 
             Ok(Property {
-                name: PropertyName::RelmContainerExtAssign,
+                name: PropertyName::RelmContainerExtAssign(span),
                 ty,
             })
         } else {

--- a/relm4-macros/src/widgets/span/property_name.rs
+++ b/relm4-macros/src/widgets/span/property_name.rs
@@ -8,7 +8,7 @@ impl Spanned for PropertyName {
         match self {
             PropertyName::Ident(ident) => ident.span(),
             PropertyName::Path(path) => path.span(),
-            PropertyName::RelmContainerExtAssign => Span2::call_site(),
+            PropertyName::RelmContainerExtAssign(span) => *span,
         }
     }
 }

--- a/relm4-macros/tests/local_ref_container.rs
+++ b/relm4-macros/tests/local_ref_container.rs
@@ -1,0 +1,45 @@
+use gtk::prelude::GtkWindowExt;
+use relm4::{gtk, ComponentParts, ComponentSender, SimpleComponent};
+
+#[derive(Default)]
+struct AppModel;
+
+#[relm4_macros::component]
+impl SimpleComponent for AppModel {
+    type Init = ();
+    type Input = ();
+    type Output = ();
+    type Widgets = AppWidgets;
+
+    view! {
+        gtk::Window {
+            set_title: Some("Simple app"),
+            set_default_width: 300,
+            set_default_height: 100,
+
+            #[local_ref]
+            my_box_ref -> gtk::Box {
+                gtk::Label {
+                    set_label: "This should compile",
+                }
+            },
+        }
+    }
+
+    fn init(
+        _init: Self::Init,
+        _root: &Self::Root,
+        _sender: ComponentSender<Self>,
+    ) -> ComponentParts<Self> {
+        let model = Self;
+
+        let my_box = gtk::Box::default();
+        let my_box_ref = &my_box;
+
+        let widgets = view_output!();
+
+        ComponentParts { model, widgets }
+    }
+
+    fn update(&mut self, _msg: (), _sender: ComponentSender<Self>) {}
+}

